codegen/src/api/metadata.rs

```diff
@@ -0,0 +1,253 @@
+            TypeDef::Variant(variant) => {
+                let mut bytes = Vec::new();
+                for var in variant.variants() {
+                    bytes.extend(self.get_variant_uid(var, set));
+                }
+                bytes
+            }
+            TypeDef::Sequence(sequence) => {
+                let mut bytes = Vec::new();
+                bytes.extend(self.get_type_uid_internal(sequence.type_param().id(), set));
+                bytes
+            }
+            TypeDef::Array(array) => {
+                let mut bytes = Vec::new();
+                bytes.extend(array.len().to_be_bytes());
+                bytes.extend(self.get_type_uid_internal(array.type_param().id(), set));
+                bytes
+            }
+            TypeDef::Tuple(tuple) => {
+                let mut bytes = Vec::new();
+                for field in tuple.fields() {
+                    bytes.extend(self.get_type_uid_internal(field.id(), set));
+                }
+                bytes
+            }
+            TypeDef::Primitive(primitive) => {
+                let mut bytes = Vec::new();
+                bytes.extend(primitive.encode());
+                bytes
+            }
+            TypeDef::Compact(compact) => {
+                let mut bytes = Vec::new();
+                bytes.extend(self.get_type_uid_internal(compact.type_param().id(), set));
+                bytes
+            }
+            TypeDef::BitSequence(bitseq) => {
+                let mut bytes = Vec::new();
+                bytes.extend(
+                    self.get_type_uid_internal(bitseq.bit_order_type().id(), set),
+                );
+                bytes.extend(
+                    self.get_type_uid_internal(bitseq.bit_store_type().id(), set),
+                );
+                bytes
+            }
+        };
+        bytes.extend(data);
+        MetadataHashable::hash(&bytes)
+    }
+
+    fn get_type_uid_internal(&self, id: u32, set: &mut HashSet<u32>) -> [u8; 32] {
+        lazy_static! {
+            static ref CACHED_UID: Mutex<HashMap<u32, [u8; 32]>> =
+                Mutex::new(HashMap::new());
+        }
+
+        if let Some(cached) = CACHED_UID.lock().unwrap().get(&id) {
+            return *cached
+        }
+
+        let ty = self.metadata.types.resolve(id).unwrap();
+
+        let mut bytes = vec![MetadataHashableIDs::Type as u8];
+        bytes.extend(ty.path().segments().concat().into_bytes());
+        // Guard against recursive types
+        if !set.insert(id) {
+            return MetadataHashable::hash(&bytes)
+        }
+
+        let ty_def = ty.type_def();
+        bytes.extend(self.get_type_def_uid(ty_def, set));
+
+        let uid = MetadataHashable::hash(&bytes);
+        CACHED_UID.lock().unwrap().insert(id, uid);
+        uid
+    }
+
+    pub fn get_type_uid(&self, id: u32) -> [u8; 32] {
+        let mut set = HashSet::<u32>::new();
+        self.get_type_uid_internal(id, &mut set)
+    }
+
+    pub fn get_pallet_uid(
```

Comment:
Bikeshedding: Personally I'd be more in favour of `_hash` than `_uid` as an ending to these function names; I initially muddled UID up in my head with the numeric ID of each type (which is also unique, just not so consistent!)